File: leet_copy_list_with_random_pointer.py
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

def pp(root):
    while root:
        print(root.val,end=" ")
        root=root.next
    print()

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return None
        newHead = Node(head.val)
        ha = {head:newHead}
        thead = head.next
        tnewHead = newHead
        while thead:
            tnode = Node(thead.val)
            tnewHead.next = tnode
            tnewHead = tnewHead.next
            ha[thead] = tnewHead
            thead = thead.next 
        thead = head
        tnewHead = newHead
        # print(t.val,newHead.val)
        # return tnewHead
        while thead and tnewHead:
            tnewHead.random = ha[thead.random] if thead.random else None
            thead = thead.next
            tnewHead = tnewHead.next
        # print(head.val,head.next.val)
        return newHead

File: LearnJava/Leet_Copy_List_with_Random_Pointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        
        Node node = head;
        
        while (node != null) {
            Node dup = new Node(node.val);
            
            dup.next = node.next;
            node.next = dup;
            
            node = dup.next;
        }
        
        node = head;
        Node dnode = null;
        while (node != null) {
            dnode = node.next;
            if (node.random != null) {
                dnode.random = node.random.next;
            }
            
            node = dnode.next;
        }
        
        node = head;
        dnode = node.next;
        Node newRoot = node.next;
        while (dnode.next != null) {
            dnode = node.next;
            node.next = dnode.next;
            dnode.next = dnode.next.next;
            
            node = node.next;
            dnode = dnode.next;
        }
        node.next = null;
        return newRoot;
    }
}